### PR TITLE
[webgpu] Merge shader_preprocessor.ts into webgpu_program.ts

### PR DESCRIPTION
**Pull request by gyagp**
The merge can solve circular import problem so that WebGPUProgram can be
removed.
isFlatDispatch is added to avoid potential inconsist changes.
yarn.lock is changed due to structual change in tfjs root folder.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6466)

### CODE REVIEW

**Comment by gyagp** [comment]
@qjia7 @xhcao PTAL

File: tfjs-backend-webgpu/src/webgpu_program.ts [hunk 1]
```diff
@@ -50,8 +50,285 @@ export interface WebGPUProgram {
...
+  } else if (index === 4) {
+    return 'u';
+  } else if (index === 5) {
+    return 'v';
+  } else {
+    throw Error(`Index ${index} is not yet supported`);
+  }
+}
+
+export function getInputByOutputSnippet(
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/webgpu_program.ts:106]: It seems that `getInputByOutputSnippet` should be a private function. `export` is not needed. Please also check other places.

File: tfjs-backend-webgpu/src/webgpu_program.ts [hunk 2]
```diff
@@ -50,8 +50,285 @@ export interface WebGPUProgram {
...
+  return [snippet, rank];
+}
+
+export function getWorkGroupSizeString(): string {
+  return `
+  @stage(compute) @workgroup_size(workGroupSizeX, workGroupSizeY, workGroupSizeZ)
+`;
 }
 
 export const makeBindGroup =
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/webgpu_program.ts:334]: After refactoring, this file is more like program and shader related. Maybe move `makeBindGroup` out of this file?

**Review by qjia7 — COMMENTED** [review]
Thanks for the refactoring.

File: tfjs-backend-webgpu/src/webgpu_program.ts (hunk 1 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/webgpu_program.ts:106]: done

File: tfjs-backend-webgpu/src/webgpu_program.ts (hunk 2 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/webgpu_program.ts]: Moved to backend_webgpu.ts

File: tfjs-backend-webgpu/src/webgpu_program.ts [hunk 3]
```diff
@@ -87,16 +67,761 @@ export const compileProgram =
...
+    globalId = GlobalId;
+    numWorkgroups = NumWorkgroups;
+`;
+}
+
+/**
+ * Generates getOutputCoords() function that computes output coordinates from
+ * dispatch geometry to reduce arithmetic.
+ */
+export function getOutputCoordsSnippet(
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/webgpu_program.ts:129]: nit:  remove `export`

**Review by qjia7 — APPROVED** [review]
LGTM with one nit, thanks.

**Pull request merged**